File: internal/detect/rule.go
package detect

import "regexp"

// Rule is one compiled detection pattern contributed by a Provider.
// Group is the regex submatch index holding the credential itself
// (0 = whole match). Severity is one of "low", "medium", "high", "critical".
type Rule struct {
	Name     string
	Severity string
	Regex    *regexp.Regexp
	Group    int
}

File: hooks/detect/rule.ts
/**
 * One detection pattern.
 *
 * `group` is the submatch index holding the credential itself (0 = the whole
 * match), so a context-bearing rule masks the secret and not the text around
 * it. Every pattern carries the `g` flag; `scan` relies on `matchAll`.
 */
export type Rule = {
  name: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  re: RegExp
  group: number
}

export const rule = (
  name: string,
  severity: Rule['severity'],
  re: RegExp,
  group = 0,
): Rule => ({ name, severity, re, group })
